classes/statistics/UsageStatsTotalTemporaryRecordDAO.inc.php

```diff
@@ -0,0 +1,369 @@
+<?php
+
+/**
+ * @file classes/statistics/UsageStatsTotalTemporaryRecordDAO.inc.php
+ *
+ * Copyright (c) 2013-2021 Simon Fraser University
+ * Copyright (c) 2003-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class UsageStatsTotalTemporaryRecordDAO
+ * @ingroup statistics
+ *
+ * @brief Operations for retrieving and adding total usage.
+ *
+ * It considers:
+ * context index page views,
+ * submission abstract, primary and supp file views,
+ * geo submission usage,
+ * COUNTER submission stats.
+ */
+
+namespace APP\statistics;
+
+use APP\core\Application;
+use Illuminate\Support\Facades\DB;
+use PKP\config\Config;
+use PKP\db\DAORegistry;
+
+class UsageStatsTotalTemporaryRecordDAO
+{
+    /**
+     * The name of the table. This table conteins all usage events.
+     */
+    public string $table = 'usage_stats_total_temporary_records';
+
+    /**
+     * Add the passed usage statistic record.
+     *
+     * @param object $entryData [
+     * 	issue_id
+     *  time
+     *  ip
+     *  canonicalURL
+     *  contextId
+     *  submissionId
+     *  representationId
+     *  assocType
+     *  assocId
+     *  fileType
+     *  userAgent
+     *  country
+     *  region
+     *  city
+     *  instituionIds
+     * ]
+     */
+    public function insert(object $entryData, int $lineNumber, string $loadId): void
+    {
+        DB::table($this->table)->insert([
+            'date' => $entryData->time,
+            'ip' => $entryData->ip,
+            'user_agent' => substr($entryData->userAgent, 0, 255),
+            'line_number' => $lineNumber,
+            'canonical_url' => $entryData->canonicalUrl,
+            'issue_id' => property_exists($entryData, 'issueId') ? $entryData->issueId : null,
+            'context_id' => $entryData->contextId,
+            'submission_id' => $entryData->submissionId,
+            'representation_id' => $entryData->representationId,
+            'assoc_type' => $entryData->assocType,
+            'assoc_id' => $entryData->assocId,
+            'file_type' => $entryData->fileType,
+            'country' => !empty($entryData->country) ? $entryData->country : '',
+            'region' => !empty($entryData->region) ? $entryData->region : '',
+            'city' => !empty($entryData->city) ? $entryData->city : '',
+            'load_id' => $loadId,
+        ]);
+    }
+
+    public function checkForeignKeys(object $entryData): array
```

Comment:
I think I would favour a different approach here, for code maintenance reasons. Areas of the code that exhaustively list out relationships tend to be badly maintained; think e.g. of the "delete journal" function, which currently goes through and manually deletes all dependent data before deleting the journal. In maintaining an entity it's easy to forget to maintain it from the journal deletion perspective, so that function is often broken or incomplete.

Performance-wise, each FK check would represent an index hit, so each would be a relatively quick query -- but yes, they do add up, especially if this is being executed on every logged event.

Because we're also logging events that might legitimately refer to objects that subsequently get deleted, I like @NateWr's suggestion of making them hardy on retrieval rather than during logging.

Reply:
I am not sure I understand 100% :-\
This function is above all needed when reprocessing a file -- I assume the current files, that are processed every day, will not have this problem -- i.e. when in the meantime an object has been deleted.
The function is called when a log file line is inserted into the temporary table (where the entries are then processed further).
I think what Nate suggests is that we use 'insert or ignore' when we then insert the data from the temporary tables into the real ones (where foreign key exists). The problem is a little bit that those inserts i.e. upserts are already complicated, but... Else, because I now use a lot of foreign keys, the insert i.e. upsert will fail when a foreign key is missing.
So, now I am not sure what do you @asmecher suggest -- shall I try to use some kind of 'insert or ignore'?

Reply:
If I understand, the problem is that the foreign keys in the log may not exist in the database, correct? And that this will increasingly cause errors up-front because we're adding formal foreign key constraints? It is good news, because we should be relying more on the database to maintain data integrity. In the case a foreign key from the log doesn't exist in the database, the best case scenario would be that a database constraint causes an error; we skip the log entry entirely and carry on to the next. Let me know if I understood correctly!

Reply:
It is true that the foreign keys in the log may not exist in the database. (Lets say a submission is deleted, but the old log file still contains it and the user wants to reprocess that log file).
How it is now: When the log file is processed line by line I check for all foreign keys in this function. If they do not exist I log the message and continue with the next line of the log file. There will be no DB failure/error and the user will get the information in the scheduled task log file.
If I should change it so that the DB constraint causes an error, I will need to see how to do it so that the processing can continue: 
Currently the log entries are first inserted into temporary tables, that do not have foreign keys. Then the data is processed there and then loaded (insert or update) into the real metrics tables, that do have foreign key constraints. So I believe if for example this function fails because of the foreign keys https://github.com/bozana/ojs/blob/6782/classes/statistics/UsageStatsTotalTemporaryRecordDAO.inc.php#L268 then it will not continue and the stats are lost.
So the only possibility is maybe to add the foreign key constraints to the temporary tables as well -- there, log file data will be inserted line by line and if one line fails it can probably continue with the next line. (The data from one log line is inserted into 4 (OMP even 6) temporary tables, so that I maybe need to take care that that failing line data is not inserted in one of them, but I think this should not be a problem).
Alec, would you say this would be a better solution? -- to have foreign key constrains in temporary tables and to let the insertion there cause an error? (I cannot provide the information about foreign key failures for the user in the scheduled task log, but they will be visible in the apache error logs).

Reply:
It's tricky, but I thiiiiink it will be best to add constraints to the temporary table as well. The more we can rely on the database to take care of data integrity in place of additional checks, the better.

Reply:
Thanks a lot! I will then try to do it so... :+1: